### PR TITLE
Hex dma2

### PR DESCRIPTION
**Pull request by vgundlur**
1. We have added the DMA Engine Pool logic in this PR
2. We have merged the master branch up to July 18th commit: 410bde74ffc238f94756726e8f6c6c045c608f1c
3. Because of merging with master there are more differences compared to the halide/hex-dma2 branch

### CODE REVIEW

File: src/runtime/hexagon_dma.cpp [hunk 1]
```diff
@@ -601,14 +606,14 @@ WEAK int halide_hexagon_dma_device_sync(void *user_context, struct halide_buffer
         << "Hexagon: halide_hexagon_dma_device_sync (user_context: " << user_context
         << " buf: " << buf << ")\n";
 
-    dma_device_handle *dev = (dma_device_handle *)buf->device;
-    halide_assert(user_context, dev->dma_engine);
+ /* dma_device_handle *dev = (dma_device_handle *)buf->device;
```
Comment (dsharletg) [inline src/runtime/hexagon_dma.cpp]: TODO on this one too?

File: src/runtime/hexagon_dma_pool.cpp [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+} hexagon_dma_virtual_engine_t;
+
+typedef struct {
+    bool used;
+    bool assigned;
+    void *engine_addr;
+} hexagon_dma_engine_t;
+
+typedef struct {
+hexagon_dma_engine_t dma_engine_list[MAX_NUMBER_OF_DMA_ENGINES];
```
Comment (dsharletg) [inline src/runtime/hexagon_dma_pool.cpp]: Indent

File: src/runtime/hexagon_dma_pool.cpp [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+#define MAX_NUMBER_OF_WORK_UNITS 4
+
+typedef struct {
+    bool in_use;
+    uint8_t num_of_engines;
+    uint8_t mapped_engines[MAX_NUMBER_OF_WORK_UNITS];
+} hexagon_dma_virtual_engine_t;
+
+typedef struct {
+    bool used;
```
Comment (dsharletg) [inline src/runtime/hexagon_dma_pool.cpp]: Comment, particularly on the difference between used vs. assigned

File: src/runtime/hexagon_dma_pool.cpp [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+            hexagon_dma_pool->dma_engine_list[i].assigned = false;
+            hexagon_dma_pool->virtual_engine_list[i].in_use = false;
+            for (int j=0; j < MAX_NUMBER_OF_WORK_UNITS; j++) {
+                hexagon_dma_pool->virtual_engine_list[i].mapped_engines[j] = 0;
+            }
+            hexagon_dma_pool->virtual_engine_list[i].num_of_engines = 0;
+        }
+    }
+
+    // halide_hexagon_max_number_of_dma_engines
```
Comment (dsharletg) [inline src/runtime/hexagon_dma_pool.cpp]: This comment probably should be removed

File: src/runtime/hexagon_dma_pool.cpp (hunk 3 above)
Comment (sfraza) [reply src/runtime/hexagon_dma_pool.cpp]: Added

File: src/runtime/hexagon_dma_pool.cpp (hunk 2 above)
Comment (sfraza) [reply src/runtime/hexagon_dma_pool.cpp]: Added indent

File: src/runtime/hexagon_dma_pool.cpp (hunk 4 above)
Comment (sfraza) [reply src/runtime/hexagon_dma_pool.cpp]: Removed

File: src/runtime/hexagon_dma.cpp (hunk 1 above)
Comment (sfraza) [reply src/runtime/hexagon_dma.cpp]: Added TODO comment . Not sure what to do in this function right now

**Pull request merged**